File: app/models/backtest.py
from datetime import date, datetime

from sqlalchemy import Date, DateTime, ForeignKey, Integer, Numeric, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class BacktestTask(Base):
    __tablename__ = "backtest_tasks"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    strategy_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("strategies.id"), nullable=True
    )
    strategy_params: Mapped[dict] = mapped_column(JSONB, default=dict)
    stock_codes: Mapped[list] = mapped_column(JSONB, default=list)
    start_date: Mapped[date] = mapped_column(Date)
    end_date: Mapped[date] = mapped_column(Date)
    initial_capital: Mapped[float] = mapped_column(Numeric(20, 2), default=1_000_000)
    status: Mapped[str] = mapped_column(String(16), default="pending")
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now(), onupdate=func.now()
    )


class BacktestResult(Base):
    __tablename__ = "backtest_results"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    task_id: Mapped[int] = mapped_column(Integer, ForeignKey("backtest_tasks.id"))
    total_return: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    annual_return: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    max_drawdown: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    sharpe_ratio: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    win_rate: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    profit_loss_ratio: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    total_trades: Mapped[int | None] = mapped_column(Integer, nullable=True)
    benchmark_return: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    alpha: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    beta: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    volatility: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    calmar_ratio: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    sortino_ratio: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    trades_json: Mapped[list] = mapped_column(JSONB, default=list)
    equity_curve_json: Mapped[list] = mapped_column(JSONB, default=list)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

File: web/src/types/backtest.ts
/** 回测执行请求 */
export interface BacktestRunRequest {
  strategy_name: string
  strategy_params?: Record<string, number | string | boolean>
  stock_codes: string[]
  start_date: string
  end_date: string
  initial_capital?: number
}

/** 回测绩效指标 */
export interface BacktestMetrics {
  total_return: number | null
  annual_return: number | null
  max_drawdown: number | null
  sharpe_ratio: number | null
  win_rate: number | null
  profit_loss_ratio: number | null
  total_trades: number
  calmar_ratio: number | null
  sortino_ratio: number | null
  volatility: number | null
  elapsed_ms?: number
}

/** 回测执行响应 */
export interface BacktestRunResponse {
  task_id: number
  status: string
  result?: BacktestMetrics | null
  error_message?: string | null
}

/** 交易记录 */
export interface TradeEntry {
  stock_code: string
  direction: string
  date: string
  price: number
  size: number
  commission: number
  pnl: number
}

/** 净值曲线数据点 */
export interface EquityCurveEntry {
  date: string
  value: number
}

/** 回测结果详情响应 */
export interface BacktestResultResponse {
  task_id: number
  status: string
  strategy_name?: string | null
  stock_codes?: string[] | null
  start_date?: string | null
  end_date?: string | null
  result?: BacktestMetrics | null
  trades?: TradeEntry[] | null
  equity_curve?: EquityCurveEntry[] | null
  error_message?: string | null
}

/** 回测任务列表项 */
export interface BacktestListItem {
  task_id: number
  strategy_name: string
  stock_count: number
  start_date: string | null
  end_date: string | null
  status: string
  annual_return: number | null
  created_at: string
}

/** 回测任务列表响应 */
export interface BacktestListResponse {
  total: number
  page: number
  page_size: number
  items: BacktestListItem[]
}
